codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ServerServiceGeneratorV2.kt

```diff
@@ -281,24 +327,21 @@ class ServerServiceGeneratorV2(
 
             impl $serviceName<()> {
                 /// Constructs a builder for [`$serviceName`].
-                pub fn builder() -> $builderName<#{NotSetGenerics:W}> {
+                /// You must specify what plugins should be applied to the operations in this service.
+                ///
+                /// Use [`$serviceName::builder_without_plugins`] if you don't need to apply plugins.
+                pub fn builder_with_plugins<Body, Plugin>(plugin: Plugin) -> $builderName<Body, Plugin> {
```

Comment:
That API would be an option. 
I think it comes with a few tradeoffs, as we were discussing privately before:

- We need to add new type to our codegen (`PokemonServicePluginBuilder`);
- We need to duplicated the setter methods across both structs (`PokemonServicePluginBuilder` and `PokemonServiceBuilder`);
- We need to keep the `Pluggable` trait around.

Using the API in this PR (and the `PluginPipeline` struct I am working on), we'd have:

```rust
// Set a plugin.
let plugin_pipeline = PluginPipeline::new().add(MyPlugin);
let app = PokemonService::builder_with_plugins(plugin_pipeline) 
    .get_pokemon_species(/* ... */) /* : PokemonServiceBuilder */
    /* ... */

// Use curated plugin stack by us
let app = PokemonService::builder_with_plugins(get_awesome_plugins()) 
    .get_pokemon_species(/* ... */) /* : PokemonServiceBuilder */
    /* ... */

// Use curated plugin stack by us plus a few of your own
let plugin_pipeline = get_awesome_plugins().add(MyPersonalPlugin);
let app = PokemonService::builder_with_plugins(plugin_pipeline) 
    .get_pokemon_species(/* ... */) /* : PokemonServiceBuilder */
    /* ... */

// Append the curated plugin stack to your own plugins, using an extensions trait
use awesome_plugins::AwesomePluginsExt;
let plugin_pipeline = PluginPipeline::new(MyPersonalPlugin).add_awesome_plugins(); 
let app = PokemonService::builder_with_plugins(plugin_pipeline) 
    .get_pokemon_species(/* ... */) /* : PokemonServiceBuilder */
    /* ... */

// Don't set a plugin.
let plugin_builder = PokemonService::builder_without_plugins() 
    .get_pokemon_species(/* ... */)
    /* ... */
```

The difference seems pretty minimal to me. We can improve it further by renaming `builder_with_plugins` to `builder` if we think that's the most common case.

Reply:
> builder_without_plugins is long enough that we may as well consider forcing users to pass the identity plugin stuck to builder().

I considered this option, but it adds a bit more burden for the user - now they need to find out that we have a no-op plugin and pass it as an argument.

Reply:
> The first two reasons constitute complexity on us.

True, but I don't think this is a reason to disregard them - when the user experience is pretty much equivalent, I think we should take into account the associated maintenance and implementation burden as one of our decision criteria.

Reply:
What about the following?
```rust
// Set a plugin.
let app = PokemonService::builder() /* : PokemonServicePluginBuilder */
    .plugin(/* ... */) /* : PokemonServiceBuilder */
    .get_pokemon_species(/* ... */) /* : PokemonServiceBuilder */
    /* ... */

// Don't set a plugin.
let plugin_builder = PokemonService::builder() /* : PokemonServicePluginBuilder */
    .without_plugins() /* : PokemonServiceBuilder */
    .get_pokemon_species(/* ... */) /* : PokemonServiceBuilder */
    /* ... */
```

It seems like a middle ground between between the current, and @hlbarber's suggested approach.
Users will see the `PokemonServicePluginBuilder` type early, which can lead them to the docs for either using/not using plugins, and we don't have to duplicate setters.

If we go for this we might want to change `.plugin()` to `.with_plugin()/.with_plugins()` for consistency.

Reply:
> What about the following?

@jjant: What benefit do you see there compared to either my proposal or @hlbarber? Is the idea that it still provides a fluent-like API without introducing confusion as to the order in which operations are allowed because you are forced to call `with_plugin`/`without_plugin` before doing anything else?

The main reason I lean against the fluent builder API proposed by @hlbarber is that it introduces a hidden piece of state.  
As a developer, I have a single `builder()` method to get my hands on a builder from `PokemonService`. 
Following the suggestions provided by the intellisense in my editor (or by looking at the docs), I can see that I have access to methods to register plugins as well as register handlers.  
If I register a handler first, suddenly the plugin methods are no longer available.
To figure out why that is the case, I need to open the docs and notice that the `plugin` method returns a different type (`PokemonServiceBuilder`) compared to the type returned by methods for registering handlers (`PokemonServiceBuilder`). 

By having `builder_with_plugins`/`builder_without_plugins`, the end-to-end journey is more obvious to the developer - the set of available options is immediately clear (i.e. you can't move forward without declaring if you want plugins or not) and they can never manage to find themselves in a state where they registered a handler but now want to add a plugin and they are not sure why that isn't possible.

At the same time, I do believe this to be a minor papercut, so if you all feel strongly about it I'm happy to refactor the API in a more fluent direction.